language/move-analyzer/src/symbols.rs

```diff
@@ -0,0 +1,2184 @@
+//! per module map is keyed on the line number where the identifier is located, and the map entry
+//! contains a list of identifier/definition pairs ordered by the column where the identifier starts.
+//!
+//! For example consider the following code fragment (0-based line numbers on the left and 0-based
+//! column numbers at the bottom):
+//!
+//! 7: const SOME_CONST: u64 = 42;
+//! 8:
+//! 9: SOME_CONST + SOME_CONST
+//!    |     |  |   | |      |
+//!    0     6  9  13 15    22
+//!
+//! Symbolication information for this code fragment would look as follows assuming that this code
+//! is stored in a file with hash FHASH (note that identifier in the definition of the constant maps
+//! to itself):
+//!
+//! [7] -> [UseLoc(7:6-13, FHASH), DefLoc(7:6, FHASH)]
+//! [9] -> [UseLoc(9:0-9 , FHASH), DefLoc((7:6, FHASH)], [UseLoc(9:13-22, FHASH), DefLoc((7:6, FHASH)]
+//!
+//! Including line number (and file hash) with the (use) identifier location may appear redundant,
+//! but it's needed to allow accumulating uses with each definition to support
+//! go-to-references. This is done in a global map from an identifier location (DefLoc) to a set of
+//! use locations (UseLoc) - we find a all references of a given identifier by first finding its
+//! definition and then using this definition as a key to the global map.
+//!
+//! Symbolication algorithm first analyzes all top-level definitions from all modules and then
+//! processes function bodies and struct definitions to match uses to definitions. For local
+//! definitions, the symbolicator builds a scope stack, entering encountered definitions and
+//! matching uses to a definition in the innermost scope.
+
+use crate::context::Context;
+use anyhow::Result;
+use codespan_reporting::files::{Files, SimpleFiles};
+use lsp_server::{Request, RequestId};
+use lsp_types::{GotoDefinitionParams, Location, Position, Range, ReferenceParams};
+use std::{
+    cmp,
+    collections::{BTreeMap, BTreeSet, VecDeque},
+    fs,
+    path::{Path, PathBuf},
+};
+use tempfile::tempdir;
+use url::Url;
+
+use move_command_line_common::files::FileHash;
+use move_compiler::{
+    diagnostics,
+    expansion::ast::{Fields, ModuleIdent, ModuleIdent_},
+    naming::ast::{StructDefinition, StructFields, TParam, Type, TypeName_, Type_},
+    parser::ast::StructName,
+    shared::Identifier,
+    typing::ast::{
+        BuiltinFunction_, Exp, ExpListItem, Function, FunctionBody_, LValue, LValueList, LValue_,
+        ModuleCall, ModuleDefinition, SequenceItem, SequenceItem_, UnannotatedExp_,
+    },
+    PASS_TYPING,
+};
+use move_ir_types::location::*;
+use move_package::compilation::build_plan::BuildPlan;
+use move_symbol_pool::Symbol;
+
+/// Enabling/disabling the language server reporting readiness to support go-to-def and
+/// go-to-references to the IDE.
+pub const DEFS_AND_REFS_SUPPORT: bool = false;
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
```

Comment:
Is there a reason not to give this `Copy`? Doesn't look like it to me

Reply:
Same for similar types below

Reply:
I did not want to add trait derivations unless the value of the type needs to be copied but does not implement the `Copy` trait and the compiler barfs.